src/plonk/circuit.rs

```diff
@@ -594,4 +606,70 @@ impl<F: Field> ConstraintSystem<F> {
         self.num_aux_columns += 1;
         tmp
     }
+
+    /// Hashes the `ConstraintSystem` into a `u64`.
+    pub fn hash(&self, mut hasher: &mut Blake2bState) -> [u8; 64] {
+        hasher.update(b"num_fixed_columns");
+        hasher.update(&self.num_fixed_columns.to_le_bytes());
+
+        hasher.update(b"num_advice_columns");
+        hasher.update(&self.num_advice_columns.to_le_bytes());
+
+        hasher.update(b"num_aux_columns");
+        hasher.update(&self.num_aux_columns.to_le_bytes());
+
+        hasher.update(b"num_gates");
+        hasher.update(&self.gates.len().to_le_bytes());
+        for gate in self.gates.iter() {
+            hasher.update(gate.0.to_owned().as_bytes());
+            gate.1.hash(&mut hasher);
+        }
+
+        hasher.update(b"num_advice_queries");
+        hasher.update(&self.advice_queries.len().to_le_bytes());
+        for query in self.advice_queries.iter() {
+            query.0.hash(&mut hasher);
+            query.1.hash(&mut hasher);
+        }
+
+        hasher.update(b"num_aux_queries");
+        hasher.update(&self.aux_queries.len().to_le_bytes());
+        for query in self.aux_queries.iter() {
+            query.0.hash(&mut hasher);
+            query.1.hash(&mut hasher);
+        }
+
+        hasher.update(b"num_fixed_queries");
+        hasher.update(&self.fixed_queries.len().to_le_bytes());
+        for query in self.fixed_queries.iter() {
+            query.0.hash(&mut hasher);
+            query.1.hash(&mut hasher);
+        }
+
+        hasher.update(b"num_permutations");
+        hasher.update(&self.permutations.len().to_le_bytes());
+        for argument in self.permutations.iter() {
+            hasher.update(&argument.get_columns().len().to_le_bytes());
+            for column in argument.get_columns().iter() {
+                column.hash(&mut hasher);
+            }
+        }
+
+        hasher.update(b"num_lookups");
+        hasher.update(&self.lookups.len().to_le_bytes());
+        for argument in self.lookups.iter() {
+            hasher.update(&argument.input_columns.len().to_le_bytes());
+            hasher.update(&argument.table_columns.len().to_le_bytes());
```

Comment:
These should be the same, you can just hash in one of them and put an assertion here that the lengths are the same.